lib/Driver/Compilation.cpp

```diff
@@ -1122,16 +1142,29 @@ static bool writeFilelistIfNecessary(const Job *job, DiagnosticEngine &diags) {
     case FilelistInfo::WhichFiles::PrimaryInputs:
       for (const Action *A : job->getSource().getInputs()) {
         const auto *IA = cast<InputAction>(A);
-        out << IA->getInputArg().getValue() << "\n";
+        const llvm::opt::Arg &InArg = IA->getInputArg();
+        if (ToolChain::canCompileInputArgumentBePrimary(job->getOutput(),
+                                                        InArg))
+          out << IA->getInputArg().getValue() << "\n";
```

Comment:
New round of comments where I've given up on improving `-index-file`: …that is, here I'd just put a method on JobContext called `isFrontendPrimaryInput` or something that takes an InputAction and checks to see if Args has `OPT_index_file_path` and if it matches up.

Reply:
(Because the *right* answer is that `-index-file` is neither a StandardCompile nor a BatchCompile nor a SingleCompile; it's a OneInputOnlyCompile or something.)

Reply:
I looked a lot at making the index-file-path handling less clever, and it gets very ugly. The problem is that the functionality is needed in several places: When constructing a command-line, when building a filelist from both `performJobImpl`, and in `addPendingJobToTaskQueue`. So one ends up passing arguments around. Then the logic to select a primary argument gets complex enough that I tried to factor it out. Then that gets hairy. I can take an other run at it tomorrow, but the clever solution has the appeal of centralizing the hack, and simplifying the downstream.

As an aside, the status quo (and this PR) duplicates the logic used to enumerate inputs and outputs between file list construction and command line creation. It used to be just inputs and outputs, but now it's inputs, primary inputs, main outputs, and supplementary outputs.
It's crying for a fix: either factoring or moving to just using filelists, as we've spoken about.

Let me know what you think or if you want to talk more offline, thanks!

Reply:
Sorry, why do we need it for `addPendingJobToTaskQueue`?

Reply:
I suspect it was added by @graydon as part of the make-batch-jobs-work-for-incremental-compilation work. It's called by `transferJobsToTaskQueue`, which is called by `formBatchJobsAndAddPendingJobsToTaskQueue`, which is called in two places: the first is likely as before: `performJobsImpl`, but the second is more interesting: it's called from `runTaskQueueToCompletion` and the comment is:
        // Task-exit callbacks from TaskQueue::execute may have unblocked jobs,
        // which means there might be PendingExecution jobs to enqueue here. If
        // there are, we need to continue trying to make progress on the
        // TaskQueue before we start marking deferred jobs as skipped, below.
So, what I deduce is going on is that batch job formation happens *after* the scheduling for incremental jobs, and that's why things are called from two places.
The idea AFAICT is to lazily do the batching so that we can more optimally schedule the compilation of individual primaries in batches.